primary/src/core.rs

```diff
@@ -263,38 +265,77 @@ impl Core {
             .inc();
 
         // Check if we can vote for this header.
-        // TODO: compute from a persisted last_voted (issue #488)
-        if self
-            .last_voted
-            .entry(header.round)
-            .or_insert_with(HashSet::new)
-            .insert(header.author.clone())
-        {
-            // Make a vote and send it to the header's creator.
-            let vote = Vote::new(header, &self.name, &mut self.signature_service).await;
-            debug!(
-                "Created vote {vote:?} for {header} at round {}",
-                header.round
-            );
-            if vote.origin == self.name {
-                // This used to be a panic, but this may fail if we're processing a late-received
-                // header where we were already are part of the signers (as a restart could have blown our `last_voted`)
-                // TODO: change this back to a panic-on-error once we have persisted last_voted (issue #488)
-                let res = self.process_vote(vote).await;
-                if let Err(e) = res {
-                    error!("Failed to process our own vote: {}", e.to_string());
+        // Send the vote when:
+        // 1. when there is no existing vote for this publicKey & round
+        //    (for this publicKey the last voted round in the votes store < header.round)
+        // 2. when there is a vote for this publicKey & round,
+        //    (for this publicKey the last voted round in the votes store = header.round)
+        //    and the hash also corresponding to the vote we already sent matches the hash of
+        //    the vote we create for this header we received
+        // Taking the inverse of these two, the only time we don't want to vote is when:
+        // there is a digest for the publicKey & round, and it does not match the digest of the
+        // vote we create for this header.
+        // Also when the header round is less than the latest round we have already voted for,
+        // then it is useless to vote, so we don't.
+
+        let result = self
+            .vote_store
+            .read(header.author.clone())
+            .await
+            .map_err(StoreError)?;
+
+        if let Some(round_digest_pair) = result {
+            if header.round < round_digest_pair.round {
+                return Ok(());
+            }
+            if round_digest_pair.round == header.round {
```

Comment:
super nit: I would keep the operands in the same order as the previous if statement `header.round == round_digest_pair.round` so could make reading even easier